pkg/events/events.go

```diff
@@ -0,0 +1,376 @@
+package events
+
+/*
+ Copyright [2019] [Platform9 Systems, Inc]
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+import (
+	"os"
+	"path"
+	"strconv"
+	"sync"
+	"time"
+
+	"k8s.io/client-go/rest"
+
+	"github.com/spf13/viper"
+
+	"k8s.io/apimachinery/pkg/watch"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	log "github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/util/workqueue"
+)
+
+const (
+	resyncPeriod   = 5 * time.Minute
+	prometheusPort = 9090
+	suffixLen      = 8
+)
+
+// Watcher watches for changes in kubernetes events
+type Watcher struct {
+	client     kubernetes.Interface
+	eventInf   cache.SharedIndexInformer
+	eventQueue workqueue.RateLimitingInterface
+	eventCache map[string]*evstore
+	lock       sync.Mutex
+}
+
+type evstore struct {
+	ev        *v1.Event
+	createdAt time.Time
+}
+
+type eventCollector struct {
+	eventMetric *prometheus.Desc
+	w           *Watcher
+}
+
+//NewEventCollector creates new collector
+func NewEventCollector(w *Watcher) *eventCollector {
+	return &eventCollector{
+		eventMetric: prometheus.NewDesc(
+			"kubernetes_events",
+			"State of kubernetes events",
+			[]string{"event_namespace", "event_name", "event_kind", "event_objname", "event_reason", "event_type", "event_message", "event_source"},
+			nil,
+		),
+		w: w,
+	}
+}
+
+func (collector *eventCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- collector.eventMetric
+}
+
+func (collector *eventCollector) Collect(ch chan<- prometheus.Metric) {
+
+	for _, e := range collector.w.eventCache {
+		ev := e.ev
+		ch <- prometheus.MustNewConstMetric(collector.eventMetric, prometheus.GaugeValue, float64(ev.Count),
+			ev.Namespace,
+			ev.Name,
+			ev.InvolvedObject.Kind,
+			ev.InvolvedObject.Name,
+			ev.Reason,
+			ev.Type,
+			ev.Message,
+			ev.Source.Component,
+		)
+	}
+}
+
+// New returns new instance of watcher
+func New() (*Watcher, error) {
```

Comment:
what's the difference between standalone and k8s? Standalone is through CLI and k8s is through qbert?

Reply:
Standalone is for debugging locally where you have a kubeconfig exported, k8s is from within the monhelper pod